### PR TITLE
Scheduler - add Scheduled#concurrentExecution() strategy

### LINKED ISSUE #9883: Missing skip- or exclusive-feature for Scheduler system
**Description**
I encountered the problem, that when the triggered method is currently processing and the same timer hits again, the method is also processed again.
Migrating from the EJB `@Schedule` annotation, this is a different behaviour. In EJB the method is processed exclusively, and during that, additional triggers are ignored.

I tried setting the method to synchronized but then - like I expected - the triggers are stacking up, firing again as soon as the method finishes.

Only current workaround is
```
  private boolean running = false;
  @Scheduled(cron = "* * * * * *")
  public void doSomethingThatSometimesTakesLonger() {
    if (running) {
      return;
    }
    running = true;
    try {
      ...Method code...
    } finally {
      running = false;
    }
  }
```
I would like to have the possibility to skip triggers on methods that are still being processed

**Implementation ideas**
Either:
- Add a configuration property
- Add a property to the `@Scheduler` annotation

### LINKED ISSUE #10203: Allow Quarkus Scheduler to handle concurrent execution
**Description**
According to the docs Quarkus Scheduler allows concurrent execution

https://quarkus.io/guides/scheduler
> The invocation of the scheduled method does not depend on the status or result of the previous invocation.

It would be nice to be able to prevent concurrent execution.

**Implementation ideas**

We can add a property to the `Scheduled` annotation that would control this. It would default to the concurrent behavior of course.

### PR DESCRIPTION
**Pull request by mkouba**
- resolves  #9883 and #10203

### CODE REVIEW

File: docs/src/main/asciidoc/scheduler-reference.adoc [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+include::./attributes.adoc[]
+
+:numbered:
+:sectnums:
+:sectnumlevels: 4
+:toc:
+
+Modern applications often need to run specific tasks periodically.
+There are two scheduler extensions in Quarkus.
+The `quarkus-scheduler` extension brings the API and and a lightweight in-memory scheduler implementation.
```
Comment (gsmet) [inline docs/src/main/asciidoc/scheduler-reference.adoc:17]:
```suggestion
The `quarkus-scheduler` extension brings the API and a lightweight in-memory scheduler implementation.
```

File: docs/src/main/asciidoc/scheduler-reference.adoc [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+NOTE: If there is no CDI scope defined on the declaring class then `@Singleton` is used. 
+
+Furthermore, the annotated method must return `void` and either declare no parameters or one parameter of type `io.quarkus.scheduler.ScheduledExecution`.
+
+TIP: The annotation is repeatable so a single method could be scheduled multiple times.
+
+=== Triggers
+
+A trigger is defined either by the `@Scheduled#cron()` or by the `@Scheduled#every()` attributes.
+If both are specified, the cron expression takes precedence.
```
Comment (gsmet) [inline docs/src/main/asciidoc/scheduler-reference.adoc:38]: Should we raise a warning in this case?

File: docs/src/main/asciidoc/scheduler-reference.adoc [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+However, if an expression starts with a digit, the `PT` prefix is added automatically.
+So for example, `15s` can be used instead of `PT15S` and is parsed as "15 seconds".
+
+[source,java]
+----
+@Scheduled(every = "2s", delayed = "2h")
+void everyTwoSeconds() { }
+----
+
+NOTE: If `@Scheduled#delay()` is set to a value greater then zero the value of `@Scheduled#delayed()` is ignored.
```
Comment (gsmet) [inline docs/src/main/asciidoc/scheduler-reference.adoc:128]: Maybe we should raise a warning too?

File: docs/src/main/asciidoc/scheduler-reference.adoc [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+    }
+}
+----
+
+== Scheduled Methods and Testing
+
+It is often desirable to disable the scheduler when running the tests.
+The scheduler can be disabled through the runtime config property `quarkus.scheduler.enabled`.
+If set to `false` the scheduler is not started even though the application contains scheduled methods.
+You can even disable scheduler for particular <<getting-started-testing#testing_different_profiles,Test Profiles>>.
```
Comment (gsmet) [inline docs/src/main/asciidoc/scheduler-reference.adoc:215]:
```suggestion
You can even disable the scheduler for particular <<getting-started-testing#testing_different_profiles,Test Profiles>>.
```

File: extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/Scheduled.java [hunk 5]
```diff
@@ -104,6 +105,14 @@
      */
     String delayed() default "";
 
+    /**
+     * Specify the strategy to handle concurrent execution of a scheduled method. By default, a scheduled method can be executed
+     * concurrently.
+     * 
+     * @return the concurrent execution strategy
+     */
+    ConcurrentExection concurrentExecution() default PROCEED;
```
Comment (gsmet) [inline extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/Scheduled.java:114]: `ConcurrentExection` -> `ConcurrentExecution`

File: extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/runtime/SkipConcurrentExecutionInvoker.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+import org.jboss.logging.Logger;
+
+import io.quarkus.scheduler.Scheduled;
+import io.quarkus.scheduler.ScheduledExecution;
+
+/**
+ * A scheduled invoker wrapper that skips concurrent executions.
+ * 
+ * @see Scheduled#concurrentExecution()
+ * @see io.quarkus.scheduler.Scheduled.ConcurrentExection#SKIP
```
Comment (gsmet) [inline extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/runtime/SkipConcurrentExecutionInvoker.java:14]: Just check the refactoring picked this one too.

File: docs/src/main/asciidoc/scheduler-reference.adoc [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+
+[[concurrent_execution]]
+=== Concurrent Execution
+
+By default, a scheduled method can be executed concurrently.
+Nevertheless, it is possible to specify the strategy to handle concurrent executions via `@Scheduled#concurrentExecution()`.
+
+[source,java]
+----
+import static io.quarkus.scheduler.Scheduled.ConcurrentExection.SKIP;
```
Comment (gsmet) [inline docs/src/main/asciidoc/scheduler-reference.adoc:148]:
```suggestion
import static io.quarkus.scheduler.Scheduled.ConcurrentExecution.SKIP;
```

**Review by gsmet — CHANGES_REQUESTED** [review]
The new doc is amazing, thanks! I spotted a few minor issues.

File: docs/src/main/asciidoc/scheduler-reference.adoc (hunk 2 above)
Comment (mkouba) [reply docs/src/main/asciidoc/scheduler-reference.adoc:38]: That's a good point.

**Comment by mkouba** [comment]
@gsmet thanks for your comments! All should be addressed now.

**Review by gsmet — APPROVED** [review]

**Pull request merged**